lib/Websocket/Rfc6455Gateway.php

```diff
@@ -573,39 +573,58 @@ private function write(Rfc6455Client $client, string $msg, int $opcode, int $rsv
         return $deferred->promise();
     }
 
-    public function send(/* int|array|null */ $clientId, string $data, bool $binary): Promise {
-        if ($clientId === null) {
-            $clientId = array_keys($this->clients);
+    public function send(int $clientId, string $data, bool $binary): Promise {
+        if (!isset($this->clients[$clientId])) {
+            return new Success;
+        }
+
+        $client = $this->clients[$clientId];
+
+        $client->messagesSent++;
+
+        $opcode = $binary ? self::OP_BIN : self::OP_TEXT;
+        assert($binary || preg_match("//u", $data), "non-binary data needs to be UTF-8 compatible");
+
+        if (\strlen($data) > 1.5 * $this->autoFrameSize) {
+            $len = \strlen($data);
+            $slices = ceil($len / $this->autoFrameSize);
+            $frames = str_split($data, (int) ceil($len / $slices));
+            $data = array_pop($frames);
+            foreach ($frames as $frame) {
+                $this->compile($client, $frame, $opcode, false);
+                $opcode = self::OP_CONT;
+            }
         }
+        return $this->compile($client, $data, $opcode);
+    }
 
-        if (\is_array($clientId)) {
+    public function broadcast(/* ?array */ $exceptIds, string $data, bool $binary): Promise {
+        if (empty($exceptIds)) {
             $promises = [];
-            foreach ($clientId as $id) {
+            foreach (array_keys($this->clients) as $id) {
```

Comment:
Isn't `foreach ($this->clients as $id => $client)` better here, as that doesn't copy the array?